### PR TITLE
Create Student Lesson Page

### PR DESCRIPTION
**Pull request by JillianK**
This creates the student lesson page at /s/<script_name>/lessons/<relative_position>/student if the script is migrated and include_student_lesson_plans is true. It has the student overview, announcements, resources, and vocab. It also has navigation to other student lesson plans (in the dropdown) and navigation to the overall script page.
Screen Shot 2021-02-18 at 9 03 02 AM
Screen Shot 2021-02-18 at 9 03 10 AM

If not migrated or include_student_lesson_plan is false:
Screen Shot 2021-02-18 at 9 14 12 AM

## Links
- [spec](https://docs.google.com/document/d/1aydi_0ZG8rLZ15UeHBYN9g_uxwK0Uq9kcNwNf0oX2t4/edit#)
-  [jira for new page](https://codedotorg.atlassian.net/browse/PLAT-315?atlOrigin=eyJpIjoiMjYzYmVlMjU2MDFiNDc3MDk3MGIzZDE5ZTJjNWJlYzQiLCJwIjoiaiJ9)
-  [jira for resources](https://codedotorg.atlassian.net/browse/PLAT-334?atlOrigin=eyJpIjoiYjY3ZmI4NjVhNDU0NDkwNGI0YjllOWQzMjczMTUwMzMiLCJwIjoiaiJ9)
- [jira for vocab](https://codedotorg.atlassian.net/browse/PLAT-317?atlOrigin=eyJpIjoiYzAxYWJkZDkxMjhkNGU3MjhkY2RmODQ3YTMwMGNhMjMiLCJwIjoiaiJ9)
- [jira for announcements](https://codedotorg.atlassian.net/browse/PLAT-338?atlOrigin=eyJpIjoiNzE0MDdkNjlhN2ZiNDdkYWFmNDg5NDY4OTBlMjgxOGMiLCJwIjoiaiJ9)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/templates/lessonOverview/StudentLessonOverview.jsx
Comment (dmcavoy) [inline apps/src/templates/lessonOverview/StudentLessonOverview.jsx]: Consider using a color from the util?

Comment (dmcavoy) [reply apps/src/templates/lessonOverview/StudentLessonOverview.jsx]: Same with backgroundColor

File: dashboard/test/controllers/lessons_controller_test.rb
Comment (dmcavoy) [inline dashboard/test/controllers/lessons_controller_test.rb:192]: Great tests check that we can't show the page in these two cases. We probably want one test to show that we can view the student lesson plan when people are supposed to

**Comment by dmcavoy** [comment]
This is looking great! I left a couple small comments but nothing major. I'll do another pass after you add the test but I think it should be ready for others to review as well. 

PS with the new PR template you don't need to keep around the sections you aren't filling in.

File: apps/src/templates/lessonOverview/StudentLessonOverview.jsx
Comment (JillianK) [reply apps/src/templates/lessonOverview/StudentLessonOverview.jsx]: oops these are actually all no longer necessary, I'll take them out

File: apps/src/templates/lessonOverview/lessonPlanShapes.jsx
Comment (dmcavoy) [inline apps/src/templates/lessonOverview/lessonPlanShapes.jsx]: I'm wondering if we want to call this studentLessonShape and then possibly even create a smaller shape for the navigation dropdown with only the critical parts called something like navigationLessonShape (I probably should have done that originally)

Comment (JillianK) [reply apps/src/templates/lessonOverview/lessonPlanShapes.jsx]: That sounds like a great idea! I'll add that

File: apps/test/unit/templates/lessonOverview/StudentLessonOverviewTest.jsx
Comment (dmcavoy) [inline apps/test/unit/templates/lessonOverview/StudentLessonOverviewTest.jsx:97]: Maybe add a test to make sure we don't show a teacher announcement?

File: apps/test/unit/templates/lessonOverview/StudentLessonOverviewTest.jsx
Comment (dmcavoy) [inline apps/test/unit/templates/lessonOverview/StudentLessonOverviewTest.jsx]: Maybe a test to make sure teacher resources don't show?

File: apps/src/templates/lessonOverview/StudentLessonOverview.jsx
Comment (dmcavoy) [inline apps/src/templates/lessonOverview/StudentLessonOverview.jsx]: Maybe we can do this on the server side to make things easier?

File: apps/test/unit/templates/lessonOverview/StudentLessonOverviewTest.jsx
Comment (JillianK) [reply apps/test/unit/templates/lessonOverview/StudentLessonOverviewTest.jsx:97]: I added the filtering into the summarize_for_student_lesson_plan like we talked about and added a test

**Review by dmcavoy — APPROVED** [review]
Great work Jillian! This came together quickly! Lets get someone else to review as well and then lets get that button added :) The curriculum and PL team are really excited to see this new feature

**Review by dmcavoy — CHANGES_REQUESTED** [review]
Can you add a UI test for the student lesson plan?

**Comment by dmcavoy** [comment]
Created this PR to pull out the ability.rb code that is not related to student lesson plans : https://github.com/code-dot-org/code-dot-org/pull/39208

File: dashboard/app/models/ability.rb
Comment (dmcavoy) [inline dashboard/app/models/ability.rb:203]: I'm currently working out the tests for viewing teacher lesson plans when they are in a pilot script. Once I have that locked down we should probably add some tests checking that this is locked to only piloters when a script is in a pilot. https://github.com/code-dot-org/code-dot-org/pull/39208

File: dashboard/app/models/ability.rb
Comment (dmcavoy) [inline dashboard/app/models/ability.rb]: I think we will want this to be just `else true end` because we don't need to check that a user in logged in to view this content

**Review by dmcavoy — APPROVED** [review]
Yay the tests are passing 🎉 ! Thank you for all your hard work to get the tests passing and update this PR. Lets get other team eyes on it

File: dashboard/app/models/lesson.rb
Comment (bethanyaconnor) [inline dashboard/app/models/lesson.rb:384]: I think you could do `all_resources.except('Teacher')` here

File: apps/src/templates/lessonOverview/StudentLessonOverview.jsx
Comment (bethanyaconnor) [inline apps/src/templates/lessonOverview/StudentLessonOverview.jsx:47]: I think it might be worth it to extract Resources display into a separate component as there is a lot of duplicate code here. I don't want to block this PR on this so I'm happy to own this as a follow-up.

**Review by bethanyaconnor — APPROVED** [review]
This looks awesome! I pulled it down locally and it's great to see. Love the tests as well. :)

I added a couple of comments, but I don't think either block this PR.

File: apps/src/sites/studio/pages/lessons/student_lesson_plan.js
Comment (davidsbailey) [inline apps/src/sites/studio/pages/lessons/student_lesson_plan.js:70]: This method looks like a duplicate: https://github.com/code-dot-org/code-dot-org/blob/4bc7fc987c9822fede43b9a32bd617487f6bf552/apps/src/sites/studio/pages/lessons/show.js#L100

It looks like it belongs in `expandableImages.js`. @Hamms does that sound right?

let's not call this blocking for merging this PR.

File: apps/src/templates/lessonOverview/StudentLessonOverview.jsx
Comment (davidsbailey) [reply apps/src/templates/lessonOverview/StudentLessonOverview.jsx:47]: @bethanyaconnor this sounds like a great next step. `linkWithQueryParams` might be worth further extracting to somewhere with other url helpers like: https://github.com/code-dot-org/code-dot-org/blob/fe313d84d94cfa95794cdc875be3ce10bcb72666/apps/src/utils.js#L720-L719 this seems like something we need to do often throughout curriculum / LP code, due to the number of different query params we carry around. `normalizeUrl` looks a bit more specific to resources.

File: dashboard/app/models/lesson.rb
Comment (JillianK) [reply dashboard/app/models/lesson.rb:384]: That gives it to me in object form aka {"Student": [], "All": []}. I thought it would be better to do the concatenation on the ruby side but I can do it on the frontend if you think that's better.

Comment (bethanyaconnor) [reply dashboard/app/models/lesson.rb:384]: Ahh makes sense. I think it's fine this way :)

File: apps/test/unit/templates/lessonOverview/StudentLessonOverviewTest.jsx
Comment (davidsbailey) [inline apps/test/unit/templates/lessonOverview/StudentLessonOverviewTest.jsx:90]: The kind of test above is problematic because it asserts the exact structure of the component, which makes it brittle. In general, we are trying to move away from this pattern.

I would propose changing this to a `mount` test, and just keeping the parts of the test which verify that values passed in as props are appearing somewhere in the rendered output (`"Lesson 1"`, `"**Algorithm**..."`, `"< Unit 1"`). I would then remove anything which expects a specific subcomponent (SafeMarkdown, InlineMarkdown, LessonNavigationDropdown) or ordering (`wrapper.find('a').at(0)`).

I know there are many examples of tests like this, so apologies for picking this PR to comment on it!

No need to block the PR on this -- I can take this one as a follow-up.

File: apps/src/templates/lessonOverview/StudentLessonOverview.jsx
Comment (JillianK) [reply apps/src/templates/lessonOverview/StudentLessonOverview.jsx:47]: I agree it feels a little silly to duplicate this and the vocab section across teacher and student lesson plans. I don't think I have the bandwidth to do that this week so making that a follow up would be great.

File: dashboard/app/controllers/lessons_controller.rb
Comment (davidsbailey) [inline dashboard/app/controllers/lessons_controller.rb]: this line should be unnecessary -- the rails convention is to render this view by default.

File: apps/src/sites/studio/pages/lessons/student_lesson_plan.js
Comment (davidsbailey) [reply apps/src/sites/studio/pages/lessons/student_lesson_plan.js:70]: Also, I looked at Course A 2021 and did not see an example of an expandable image. does anyone have a link to an example of one of these on a student lesson plan? just checking that this functionality is needed here.

**Review by davidsbailey — APPROVED** [review]
Awesome job stringing this all together, @JillianK ! I pulled it down locally and enabled lesson plans on Course A, and it looks great!

File: apps/src/sites/studio/pages/lessons/student_lesson_plan.js
Comment (dmcavoy) [reply apps/src/sites/studio/pages/lessons/student_lesson_plan.js:70]: I don't think there are examples of this right now but we support adding images to the student overview field on the edit page so it seems like we need to prepare for that to be a possibility soon?

File: dashboard/app/models/ability.rb
Comment (dmcavoy) [reply dashboard/app/models/ability.rb:203]: Just following up. I will own adding tests in my PR.

File: apps/src/sites/studio/pages/lessons/student_lesson_plan.js
Comment (davidsbailey) [reply apps/src/sites/studio/pages/lessons/student_lesson_plan.js:70]: OK. @JillianK have you tested expandable images locally? I would suggest a manual test before merging, if that is not too hard to do.

Comment (JillianK) [reply apps/src/sites/studio/pages/lessons/student_lesson_plan.js:70]: I have not, I just left this in there because I thought we would need it at some point. I can try to do that

Comment (dmcavoy) [reply apps/src/sites/studio/pages/lessons/student_lesson_plan.js:70]: Pushed an update that makes expandable images work fully!

**Pull request merged**
